### PR TITLE
macOS Support

### PR DESCRIPTION
**Pull request by safaiyeh**
## Description

Added support for macOS. This is the beginning of macOS support. Currently did enough to get the build to work on macOS, I'll spend the next couple of days making sure the returned values are correct!

I do not mind PRs to help out 😁

~~One thing I noticed is now the `iOS` folder has implementations for iOS, tvOS, and macOS. A common scheme Microsoft uses and react-native-webview uses is naming the directory `apple`, should we follow that scheme?~~ Changed to apple 👍

Possible functions that need more expertise to be implemented or more time outside the scope of this:
getBatteryLevel(): https://developer.apple.com/documentation/kernel/iopmpowersource#//apple_ref/doc/uid/TP40012438
isCameraPresent(): https://developer.apple.com/documentation/avfoundation/avcapturedevice
getInstallerPackageName(): is this the same as getBundleId()?

## Check List before ready for merge
- [ ] Update documentation
- [ ] GitHub Action CI similar to Windows implementation
- [ ] Go through each method, determine if possible for macOS & what it should return (updating table as progress is made)
 
| Method                                                            | Return Type         |   Possible? | Return value 
| ----------------------------------------------------------------- | ------------------- |  :-: |  :-: |
| [getAndroidId()](#getandroidid)                                   | `Promise<string>`    | ❌  | unknown
| [getApiLevel()](#getapilevel)                                     | `Promise<number>`   | ❌ | -1
| [getApplicationName()](#getapplicationname)                       | `string`          | ✅ | DeviceInfoTest 
| [getAvailableLocationProviders()](#getAvailableLocationProviders) | `Promise<Object>`   | ✅ |   `{ headingAvailable: false, locationServicesEnabled: true, significantLocationChangeMonitoringAvailable: true }`
| [getBaseOs()](#getbaseOs)                                         | `Promise<string>`   | ✅ | macOS
| [getBuildId()](#getbuildid)                                       | `Promise<string>`   | ✅ | 19G73
| [getBatteryLevel()](#getbatterylevel)                             | `Promise<number>`   
| [getBootloader()](#getbootloader)                                 | `Promise<string>` | ❌ | unknown
| [getBrand()](#getbrand)                                           | `string`            | ✅ | Apple
| [getBuildNumber()](#getbuildnumber)                               | `string` | ✅ | 1           
| [getBundleId()](#getbundleid)                                     | `string` | ✅ | org.reactjs.native.DeviceInfoTest          
| [isCameraPresent()](#iscamerapresent)                             | `Promise<boolean>`  
| [getCarrier()](#getcarrier)                                       | `Promise<string>`   | ❌ | unknown
| [getCodename()](#getcodename)                                     | `Promise<string>` | ❌ | unknown
| [getDevice()](#getdevice)                                         | `Promise<string>` | ❌ | unknown   
| [getDeviceId()](#getdeviceid)                                     | `string` | ✅ | Macmini8,1        
| [getDeviceType()](#getDeviceType)                                 | `string` | ✅ | Mac            
| [getDisplay()](#getdisplay)                                       | `Promise<string>` | ❌ | unknown
| [getDeviceName()](#getdevicename)                                 | `Promise<string>` | ✅ | Jasons-Mac-Mini   
| [getDeviceToken()](#getdevicetoken)                               | `Promise<string>`   | ✅ | `[Error: The operation couldn’t be completed. (com.apple.devicecheck.error error 0.)]`
| [getFirstInstallTime()](#getfirstinstalltime)                     | `Promise<number>` | ❌ | -1  
| [getFingerprint()](#getfingerprint)                               | `Promise<string>` | ❌ | unknown
| [getFontScale()](#getfontscale)                                   | `Promise<number>` | ❌ | -1
| [getFreeDiskStorage()](#getfreediskstorage)                       | `Promise<number>` | ✅ | 100742909952   
| [getHardware()](#gethardware)                                     | `Promise<string>` | ❌ | unknown   
| [getHost()](#gethost)                                             | `Promise<string>` | ❌ | unknown  
| [getIpAddress()](#getipaddress)                                   | `Promise<string>` | ✅ | 192.xxx.xxx.xx (redacted)  
| [getIncremental()](#getincremental)                               | `Promise<string>` | ❌ | unknown  
| [getInstallerPackageName()](#getinstallerpackagename)             | `Promise<string>`   
| [getInstallReferrer()](#getinstallreferrer)                       | `Promise<string>`|  ❌ | unknown
| [getInstanceId()](#getinstanceid)                                 | `Promise<string>`  | ❌ | unknown 
| [getLastUpdateTime()](#getlastupdatetime)                         | `Promise<number>` | ❌ | -1  
| [getMacAddress()](#getmacaddress)                                 | `Promise<string>` | ❌ | unknown   
| [getManufacturer()](#getmanufacturer)                             | `Promise<string>` | ✅ | Apple   
| [getMaxMemory()](#getmaxmemory)                                   | `Promise<number>` | ❌ | -1   
| [getModel()](#getmodel)                                           | `string` | ✅ | Mac mini (2018)       
| [getPhoneNumber()](#getphonenumber)                               | `Promise<string>` | ❌ | unknown
| [getPowerState()](#getpowerstate)                                 | `Promise<object>`   
| [getProduct()](#getproduct)                                       | `Promise<string>`   
| [getPreviewSdkInt()](#getPreviewSdkInt)                           | `Promise<number>`   
| [getReadableVersion()](#getreadableversion)                       | `string`            
| [getSerialNumber()](#getserialnumber)                             | `Promise<string>`   
| [getSecurityPatch()](#getsecuritypatch)                           | `Promise<string>`   
| [getSystemAvailableFeatures()](#getSystemAvailableFeatures)       | `Promise<string[]>` 
| [getSystemName()](#getsystemname)                                 | `string`            
| [getSystemVersion()](#getsystemversion)                           | `string`            
| [getTags()](#gettags)                                             | `Promise<string>`  
| [getType()](#gettype)                                             | `Promise<string>`   
| [getTotalDiskCapacity()](#gettotaldiskcapacity)                   | `Promise<number>`   
| [getTotalMemory()](#gettotalmemory)                               | `Promise<number>`   
| [getUniqueId()](#getuniqueid)                                     | `string`            
| [getUsedMemory()](#getusedmemory)                                 | `Promise<number>`   
| [getUserAgent()](#getuseragent)                                   | `Promise<string>`   
| [getVersion()](#getversion)                                       | `string`            
| [hasNotch()](#hasNotch)                                           | `boolean`          
| [hasSystemFeature()](#hassystemfeaturefeature)                    | `Promise<boolean>`  
| [isAirplaneMode()](#isairplanemode)                               | `Promise<boolean>`  
| [isBatteryCharging()](#isbatterycharging)                         | `Promise<boolean>`  
| [isEmulator()](#isemulator)                                       | `Promise<boolean>`  
| [isLandscape()](#isLandscape)                                     | `Promise<boolean>` 
| [isLocationEnabled()](#isLocationEnabled)                         | `Promise<boolean>` 
| [isHeadphonesConnected()](#isHeadphonesConnected)                 | `Promise<boolean>`  
| [isPinOrFingerprintSet()](#ispinorfingerprintset)                 | `Promise<boolean>`  
| [isTablet()](#istablet)                                           | `boolean`           
| [supported32BitAbis()](#supported32BitAbis)                       | `Promise<string[]>` 
| [supported64BitAbis()](#supported64BitAbis)                       | `Promise<string[]>` 
| [supportedAbis()](#supportedAbis)                                 | `Promise<string[]>` 
| [syncUniqueId()](#syncuniqueid)                                   | `Promise<string>`   
---
## Compatibility

| OS      | Implemented |
| ------- | :---------: |
| iOS     |    ❌     |
| Android |    ❌     |
| Windows |    ❌     |
| macOS |    ✅     |

## Checklist

* [ ] I have tested this on a device/simulator for each compatible OS
* [ ] I added the documentation in `README.md`
* [ ] I mentioned this change in `CHANGELOG.md`
* [ ] I updated the typings files (`privateTypes.ts`, `types.ts`)
* [ ] I added a sample use of the API (`example/App.js`)

### CODE REVIEW

**Comment by kaiyes** [comment]
I felt that folder structure is pretty intuitive. You can go to each folder and do things there separately without affecting other ones. It has become a second habit for me to go the mac folder and run - `mac` which is an alias for `npx react-native run-macos`.

File: ios/RNDeviceInfo/RNDeviceInfo.m
Comment (mikehardy) [inline ios/RNDeviceInfo/RNDeviceInfo.m:75]: I recognize it's draft and you're just getting compile to work, but seems OSX should likely have battery information as well? Laptop use case?

File: ios/RNDeviceInfo/RNDeviceInfo.m
Comment (mikehardy) [inline ios/RNDeviceInfo/RNDeviceInfo.m:115]: if TARGET_OS_OSX DeviceTypeComputer ? Or DeviceTypeMac ? something like that would work for me

File: ios/RNDeviceInfo/RNDeviceInfo.m
Comment (mikehardy) [inline ios/RNDeviceInfo/RNDeviceInfo.m:127]: what is log? (I'm learning it all myself)

File: ios/RNDeviceInfo/RNDeviceInfo.m
Comment (mikehardy) [inline ios/RNDeviceInfo/RNDeviceInfo.m:769]: Seems like this would be better restructured as a series instead of branching.
Initialization with locationServicesEnabled
if OSX or iOS add significantLocationChangeMonitorAvailable and headingAvailable
if iOS add isRangingAvailable

File: ios/RNDeviceInfo/RNDeviceInfo.m
Comment (safaiyeh) [reply ios/RNDeviceInfo/RNDeviceInfo.m:115]: Sounds good, I'll do DeviceTypeMac

File: ios/RNDeviceInfo/RNDeviceInfo.m
Comment (safaiyeh) [reply ios/RNDeviceInfo/RNDeviceInfo.m:127]: Hm interesting, random string on my end but didn't break the build. Ill remove

File: ios/RNDeviceInfo/RNDeviceInfo.m
Comment (safaiyeh) [reply ios/RNDeviceInfo/RNDeviceInfo.m:75]: Yup, still need to implement that.

Documented here: https://developer.apple.com/documentation/kernel/iopmpowersource#//apple_ref/doc/uid/TP40012438

File: src/internal/nativeInterface.ts
Comment (mikehardy) [inline src/internal/nativeInterface.ts:24]:
```suggestion
  • If you are using CocoaPods for apple targets, run \`pod install\` in the \`macos\` directory, then rebuild and re-run the app. You may also need to re-open Xcode to get the new pods.
```

I think this is more accurate? I'm always iffy on wording and documentation though, feel free to mash the words around, but I don't think the ios directory will exist so that at least should be removed

**Comment by mikehardy** [comment]
You are steadily chewing through this! nice

**Comment by safaiyeh** [comment]
Yup! Aiming to finish this week 😁

**Comment by mikehardy** [comment]
Excellent! I'm not making any promises (open source and all...) but I've been reading about github actions and I hope to have some CI up including a catalyst and macos build target CI check so that none of us accidentally silently break your work in the future :)

**Comment by safaiyeh** [comment]
Sounds good! I can take a stab at it too after I finish up the functions

**Comment by mezod** [comment]
Any updates on this? :D

**Comment by mikehardy** [comment]
@mezod I imagine the branch is already fit for consumption, it just doesn't implement everything it could. How did it go when you pulled the branch tried it?

**Comment by mezod** [comment]
We haven't tried it yet, we were waiting for the merge, but it's true we could just point to the fork, thanks for the prompt response :)

**Comment by safaiyeh** [comment]
We could either merge this now and let the community bring support for the rest

Or wait til they are all finished, I've been off and on on this.

What do you think? @mikehardy

**Comment by mikehardy** [comment]
You are the current 'motor' here, as long as it runs and the apis are correctly marked as supported or not I am fine merging and doing anything else as a follow up

**Comment by safaiyeh** [comment]
Okay sounds good! Run a test and create an issue for what is missing.

**Comment by kyle-ssg** [comment]
Had a quick test, compilation errors due to 
image

**Comment by mikehardy** [comment]
Thanks for checking it @kyle-ssg ! Also it has conflicts now :sweat_smile: not sure how @safaiyeh feels about it but it has sat a bit - if you wanted to clone his branch and fix it up and submit it I'm happy to take PRs etc as long as we're all working together to move it forward

**Comment by kyle-ssg** [comment]
No problem @mikehardy, I'm looking into how good RN OSX is as a whole for future projects, I imagine forking / fixing bits won't be too tricky - will confirm based on how things go with other libs.

**Comment by safaiyeh** [comment]
Hello! Sorry about forgetting about this, @mikehardy do you have a checklist needed for this get in?

**Comment by mikehardy** [comment]
Well, it shouldn't have conflicts :-) - it currently has two but I'm sure they are small, and in general you should attest that it works - nothing too crazy since it's new support. I'll check things still work on android and ios then we go for it.

**Comment by safaiyeh** [comment]
Sounds good! I'll check out the conflicts. Then open the PR

**Comment by safaiyeh** [comment]
@mikehardy Resolved the conflicts and opened the PR. Let me know any issues that come up

**Comment by mikehardy** [comment]
I should have merged this before merging something else that will conflict, I'll own that and get this merged

File: apple/RNDeviceInfo/RNDeviceInfo.m
Comment (mikehardy) [inline apple/RNDeviceInfo/RNDeviceInfo.m:32]: I'm going to handle this one differently - #1137 added ability to return Desktop as a device type, and that is vendor-neutral - I want to return that for macos platform builds

The combo of Platform === iOS and DeviceType === Desktop === Catalyst builds
The combo of Platform === macOS and DeviceType === Desktop will === Desktop as well

**Review by mikehardy — DISMISSED** [review]
These look pretty good in general and I apologize for not merging them yet
There are conflicts now but as mentioned separately I have to own that for not merging it sooner, I'll get this merged shortly with a couple patch ups

@safaiyeh can you point me to the definitive guide you would use in order to create a react-native macos from a template? I assume it is https://microsoft.github.io/react-native-windows/ (from the microsoft/react-native-macos repo) but want to make sure

Thanks @safaiyeh

**Comment by safaiyeh** [comment]
> These look pretty good in general and I apologize for not merging them yet
> There are conflicts now but as mentioned separately I have to own that for not merging it sooner, I'll get this merged shortly with a couple patch ups
> 
> @safaiyeh can you point me to the definitive guide you would use in order to create a react-native macos from a template? I assume it is https://microsoft.github.io/react-native-windows/ (from the microsoft/react-native-macos repo) but want to make sure
> 
> Thanks @safaiyeh

Yup that would be the way to go for the macOS template!

Sounds good on the conflicts - let me know if I could help.

**Comment by sraka1** [comment]
@mikehardy any approximate ETA for this being merged or should I just use safaiyeh's fork for now? Thanks :)

**Comment by kyle-ssg** [comment]
If im not mistaken, this is now released with v7.4.0 and this pull can be closed?

**Comment by mikehardy** [comment]
This is not merged the work is / would be all done here, I can't give an ETA, it's unpaid work so I prioritize paying items first and I have a few of those at the moment.

Definitely use the fork if macos is a current need, forward porting the or itself so it is merge ready would win my entire quota of internet points for the week!

File: apple/RNDeviceInfo/RNDeviceInfo.m
Comment (safaiyeh) [reply apple/RNDeviceInfo/RNDeviceInfo.m:32]: Sounds good!

**Comment by weraaaaz** [comment]
What's the status of this one?

**Comment by mikehardy** [comment]
Exactly as commented above

**Comment by mikehardy** [comment]
Still on me but this library is active, just not this branch. I do have github sponsors available if this is a business need for anyone, we can chat, otherwise this is on the "unpaid work" pile and will struggle for priority

**Comment by ShivamJoker** [comment]
@mikehardy it's now the end of 2021 🥲

**Comment by mikehardy** [comment]
Yes indeed, and in the end I've had no time for this. Still something I'd like to do and it's on my "important not urgent" list. Unfortunately "important and urgent" has had a deep queue entire second half of the year. Perhaps in 2022.

**Comment by mikehardy** [comment]
maintainer pinging themselves ;-)

**Comment by mikehardy** [comment]
still on my list, just contract work killin' me at the moment

**Comment by mikehardy** [comment]
I continue to be sad I haven't had the time to get this in yet. Some day...

**Comment by elmcapp** [comment]
Any updates as of Sep 1st 2022. is there support for macOS yet

**Comment by mikehardy** [comment]
No sorry, this is eternally back-burner. It's something I want to do but I just don't have the time and the time I do have is better spent on higher value (that is: impacts more users) projects, like getting macCatalyst to even *work* with react-native, which I did this Spring.

I'd love for anyone to take this up, and I'm sorry I could get it done but that's living...

**Comment by elmcapp** [comment]
Where you able to get macCatalyst  working with React Native. if so do you have a GitHub link to that project. I have been looking to use macCatalyst  with React Native as well

**Comment by mikehardy** [comment]
@elmcapp I have a fully scripted demonstration (you'll need to set your DEVELOPMENT_TEAM_ID) though, with react-native-firebase working, on macCatalyst, here: https://github.com/mikehardy/rnfbdemo/blob/main/make-demo.sh

In particular there used to be a number of workarounds involved:

https://github.com/mikehardy/rnfbdemo/blob/058dc304ca739f40698225753c12164875af1907/make-demo.sh#L249-L282

but myself and a few others got them upstream in a single flag now - https://github.com/facebook/react-native/blob/12e5842e11ecd53755f43e99c592135b58ac694b/template/ios/Podfile#L37-L39

Those will release with react-native 0.70

**Comment by fi-nik** [comment]
Hey @mikehardy, i'm willing to give it a shot, try to solve this one out. As i can see, it's required to support all methods, which could be possible on macOS?

**Comment by mikehardy** [comment]
@fi-nik that is the general idea - support all the methods that could be possible - but incremental progress (supporting the methods that are important to you) would be a fine start

**Comment by fi-nik** [comment]
> @fi-nik that is the general idea - support all the methods that could be possible - but incremental progress (supporting the methods that are important to you) would be a fine start
@mikehardy 
I have created draft PR for support. https://github.com/react-native-device-info/react-native-device-info/pull/1490
Everything is working on sample app, but i just need to see if any of those with ❓ can be implemented, and if they can, what are those with highest priority for now

**Pull request closed without merging**
